app/javascript/components/batch_update/components/BatchUpdateForm.js

```diff
@@ -140,11 +178,89 @@ class BatchUpdateForm extends React.Component {
     })
   }
 
+  onAddTag({ name: tag }) {
+    const { tagsToAdd } = this.state
+    this.setState({
+      tagsToAdd: [...tagsToAdd, tag],
+    })
+  }
+
+  onRemoveExistingTag(name) {
+    const tag = name
+    const { tagsToRemove } = this.state
+    this.setState({
+      tagsToRemove: [...tagsToRemove, tag],
+    })
+  }
+
+  onCancelAddTag(name) {
+    const tag = name.substr(1)
+    const { tagsToAdd } = this.state
+    this.setState({
+      tagsToAdd: tagsToAdd.filter(t => t !== tag),
+    })
+  }
+
+  onCancelRemoveTag(name) {
+    const tag = name.substr(1)
+    const { tagsToRemove } = this.state
+    this.setState({
+      tagsToRemove: tagsToRemove.filter(t => t !== tag),
+    })
+  }
+
+  getTagState() {
+    const { existingTags, tagsToAdd, tagsToRemove } = this.state
+    const output = existingTags.map(tag => ({
+      tag,
+      toAdd: false,
+      toRemove: false,
+    }))
```

Comment:
Seeing how this all gets turned into a key/value structure eventually strengthens my hunch about treating as a dictionary from the get go. I think we might be able to express the same idea more succinctly...